lottery/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+    /// * List of players must not be empty
+    async fn pick_winner(&mut self) {
+        if msg::source() == self.lottery_owner
+            && self.lottery_state.lottery_started
+            && self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration
+                <= exec::block_timestamp()
+            && !self.players.is_empty()
+        {
+            let index = (self.get_random_number() % (self.players.len() as u32)) as usize;
+            let win_player_index = *self.players.keys().nth(index).expect("Player not found");
+            let player = self.players[&win_player_index];
+
+            if self.token_address.is_some() {
+                debug!("Transfer tokens to the winner");
+                self.transfer_tokens(&exec::program_id(), &player.player_id, self.lottery_balance)
+                    .await;
+
+                self.lottery_balance = 0;
+            } else {
+                msg::send_bytes(player.player_id, b"Winner", exec::value_available());
+            }
+
+            self.lottery_history
+                .insert(self.lottery_id, player.player_id);
+            msg::reply(Event::Winner(win_player_index), 0);
+
+            debug!(
+                "Winner: {} token_address(): {:?}",
+                index, self.token_address
+            );
+
+            self.token_address = None;
+            self.lottery_state = LotteryState::default();
+            self.players = BTreeMap::new();
+        } else {
+            panic!(
+                "pick_winner(): Owner message: {}  lottery_duration: {}  players.is_empty(): {}",
+                msg::source() == self.lottery_owner,
+                self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration
+                    > exec::block_timestamp(),
+                self.players.is_empty()
+            );
+        }
+    }
+}
+
+static mut LOTTERY: Option<Lottery> = None;
+
+#[gstd::async_main]
+async unsafe fn main() {
+    if msg::source() == ZERO_ID {
+        panic!("Message from zero address");
+    }
+
+    let action: Action = msg::load().expect("Could not load Action");
+    let lottery: &mut Lottery = unsafe { LOTTERY.get_or_insert(Lottery::default()) };
+
+    match action {
+        Action::Enter(amount) => {
+            lottery.enter(amount).await;
+        }
+
+        Action::StartLottery {
+            duration,
+            token_address,
+        } => {
+            lottery.start_lottery(duration, token_address);
+        }
+
+        Action::LotteryState => {
+            msg::reply(Event::LotteryState(lottery.lottery_state), 0);
+            debug!("LotteryState: {:?}", lottery.lottery_state);
+        }
+
+        Action::PickWinner => {
+            lottery.pick_winner().await;
+        }
+
+        Action::BalanceOf(index) => {
+            lottery.get_balance(index);
+        }
+
+        Action::GetPlayers => {
+            lottery.get_players();
+        }
+
+        Action::LeaveLottery(index) => {
```

Comment:
Is it convenient for the user to point to its index? Maybe it is better to find it by `msg::source()`?
```suggestion
        Action::LeaveLottery => {
```

Reply:
Searching by key in BTreeMap<u32, Player> seemed to be the best option in terms of performance